include/oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h

```diff
@@ -35,25 +36,86 @@ class sycl_backend
     using execution_resource_t = resource_type;
     using resource_container_t = std::vector<execution_resource_t>;
 
+    bool has_enable_profiling = false;
   private:
-    class async_waiter
+    class async_waiter_base{
+        public:
+            virtual void wait() = 0;
+            virtual void report() = 0;
+            virtual bool is_complete() = 0;
+            virtual ~async_waiter_base() = default;
+    };
+
+    template<typename Selection>
+    class async_waiter : public async_waiter_base
     {
         sycl::event e_;
-
+        std::shared_ptr<Selection> s;
       public:
-        async_waiter(const sycl::event& e) : e_(e) {}
+        async_waiter(sycl::event e) : e_(e) {}
+        async_waiter(sycl::event e, std::shared_ptr<Selection> selection) : e_(e), s(selection) {}
+
+        async_waiter(async_waiter &w) : e_(w.e_), s(w.s) {}
+        async_waiter& operator=(async_waiter &w){
+            s = w.s;
+            e_ = w.e_;
+            return *this;
+        }
         sycl::event
         unwrap()
         {
             return e_;
         }
+
         void
-        wait()
+        wait() override
         {
             e_.wait();
         }
+
+        void
+        report() override{
+            if constexpr (report_value_v<Selection, execution_info::task_time_t>){
+                cl_ulong time_start = e_.template get_profiling_info<sycl::info::event_profiling::command_start>();
+                cl_ulong time_end = e_.template get_profiling_info<sycl::info::event_profiling::command_end>();
+                s->report(execution_info::task_time, time_end-time_start);
+            }
+
+        }
```

Comment:
This implementation is very problematic.
A small problem that it uses `cl_ulong` for the timestamps, but `event::get_profiling_info<>` returns `uint64_t` according to the spec.

The big problem is that the timestamps are essentially implementation-defined in SYCL:
> The possible values for each info parameter and any restrictions are defined in the specification of the SYCL backend associated with the event.

Moreover, the implementation of the policy allows using queues with and without event profiling; in the latter case, the timestamps are taken on the host. All that means that the difference between two timestamps cannot be used to compare execution time for different queues without first being somehow converted to the same units. 

I think this is the showstopper for the patch.

Reply:
Looks like I am mistaken here. Reading the SYCL spec further, I see that these timestamps are required to be in nanoseconds and have the same time base per platform.
But let's make it well documented in the code, at least.

Reply:
[`https://en.cppreference.com/w/cpp/header/chrono](https://en.cppreference.com/w/cpp/header/chrono) provide good solutions for that.
For example we may use one of `std::chrono::nanoseconds`, `std::chrono::microseconds` or `std::chrono::milliseconds` in all this internal API.

Reply:
Fixed. Thank you @SergeyKopienko